packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts

```diff
@@ -400,11 +398,10 @@ export class Pipeline extends PipelineBase {
           `got '${action.actionProperties.owner}' (Action: '${action.actionProperties.actionName}' in Stage: '${stage.stageName}')`);
       }
       actionRole = action.actionProperties.role;
-    } else if (action.actionProperties.resource) {
-      const resourceStack = Stack.of(action.actionProperties.resource);
-      // check if resource is from a different account
-      if (pipelineStack.environment !== resourceStack.environment) {
-        // if it is, the pipeline's bucket must have a KMS key
+    } else {
```

Comment:
Can you simplify the logic here? Nested "if"s make it really hard to ensure that this is actually going to end up in a desired state. Try to rewrite in "early bailout" style:

```
if (something) {
  do;
  return;
}

if (something and  something or something) {
  do;
  return;
}

//...
```

Reply:
Add more code commentary to explain what you are doing and why before each case

Reply:
I think that's virtually impossible for this function. You see, the `if`-`else` chain is done before the following code:

```ts
    if (!actionRole && this.isAwsOwned(action)) {
      // generate a Role for this specific Action
      actionRole = new iam.Role(actionScope, 'CodePipelineActionRole', {
        assumedBy: new iam.AccountPrincipal(pipelineStack.account),
      });
    }

    // the pipeline role needs assumeRole permissions to the action role
    if (actionRole) {
      this.role.addToPolicy(new iam.PolicyStatement({
        actions: ['sts:AssumeRole'],
        resources: [actionRole.roleArn]
      }));
    }
```

Which is meant to be shared between all branches of the `if`-`else` above. So I'm not sure how would guard clauses work in this case.

Reply:
See my proposal below as an example on how to think about this code in piecemeal logical conditions that require less cognitive load to digest.

Reply:
Refactored to reduce nesting. Let me know if you like this way better.